File: pulse-kotlin/src/commonMain/kotlin/com/fluxchess/pulse/kotlin/engine/Position.kt
/*
 * Copyright 2013-2024 Phokham Nonava
 *
 * Use of this source code is governed by the MIT license that can be
 * found in the LICENSE file.
 */

package com.fluxchess.pulse.kotlin.engine

class Position {
	var activeColor: Color = NO_COLOR
	var castlingRights: Castling = NO_CASTLING
	var enPassantSquare: Square = NO_SQUARE
	var halfmoveClock: Int = 0
	var halfmoveNumber: Int = 0

	internal val board: Array<Piece> = Array(SQUARES_MAX_VALUE) { NO_PIECE }
	internal val pieces: Array<BitboardArray> = Array(colors.size) { BitboardArray(pieceTypes.size) { 0u } }

	private val stateList: StateList = StateList()

	fun makeMove(move: Move) {
		saveState()

		val moveType = moveTypeOf(move)
		val originSquare = originSquareOf(move)
		val targetSquare = targetSquareOf(move)
		val originPiece = originPieceOf(move)
		val originColor = pieceColorOf(originPiece)
		val targetPiece = targetPieceOf(move)

		if (targetPiece != NO_PIECE) {
			var captureSquare = targetSquare
			if (moveType == EN_PASSANT_MOVE) {
				val oppositeDirection = pawnMoveDirections[oppositeOf(originColor)]
				captureSquare += oppositeDirection
			}
			remove(captureSquare)
			clearCastling(captureSquare)
		}

		remove(originSquare)
		if (moveType == PAWN_PROMOTION_MOVE) {
			put(pieceOf(originColor, promotionOf(move)), targetSquare)
		} else {
			put(originPiece, targetSquare)
		}

		if (moveType == CASTLING_MOVE) {
			when (targetSquare) {
				G1 -> put(remove(H1), F1)
				C1 -> put(remove(A1), D1)
				G8 -> put(remove(H8), F8)
				C8 -> put(remove(A8), D8)
				else -> error("Invalid target square: $targetSquare")
			}
		}

		clearCastling(originSquare)

		if (moveType == PAWN_DOUBLE_MOVE) {
			val oppositeDirection = pawnMoveDirections[oppositeOf(originColor)]
			enPassantSquare = targetSquare + oppositeDirection
		} else {
			enPassantSquare = NO_SQUARE
		}

		activeColor = oppositeOf(activeColor)

		if (pieceTypeOf(originPiece) == PAWN || targetPiece != NO_PIECE) {
			halfmoveClock = 0
		} else {
			halfmoveClock++
		}

		halfmoveNumber++
	}

	fun undoMove(move: Move) {
		val moveType = moveTypeOf(move)
		val originSquare = originSquareOf(move)
		val targetSquare = targetSquareOf(move)
		val originPiece = originPieceOf(move)
		val originColor = pieceColorOf(originPiece)
		val targetPiece = targetPieceOf(move)

		halfmoveNumber--

		activeColor = oppositeOf(activeColor)

		if (moveType == CASTLING_MOVE) {
			when (targetSquare) {
				G1 -> put(remove(F1), H1)
				C1 -> put(remove(D1), A1)
				G8 -> put(remove(F8), H8)
				C8 -> put(remove(D8), A8)
				else -> error("Invalid target square: $targetSquare")
			}
		}

		remove(targetSquare)
		put(originPiece, originSquare)

		if (targetPiece != NO_PIECE) {
			var captureSquare = targetSquare
			if (moveType == EN_PASSANT_MOVE) {
				val oppositeDirection = pawnMoveDirections[oppositeOf(originColor)]
				captureSquare += oppositeDirection
			}
			put(targetPiece, captureSquare)
		}

		restoreState()
	}

	private fun saveState() {
		val entry = stateList.entries[stateList.size]
		entry.castlingRights = castlingRights
		entry.enPassantSquare = enPassantSquare
		entry.halfmoveClock = halfmoveClock
		stateList.size++
	}

	private fun restoreState() {
		stateList.size--
		val entry = stateList.entries[stateList.size]
		castlingRights = entry.castlingRights
		enPassantSquare = entry.enPassantSquare
		halfmoveClock = entry.halfmoveClock
	}

	fun get(square: Square): Piece {
		return board[square]
	}

	fun put(piece: Piece, square: Square) {
		val color = pieceColorOf(piece)
		val pieceType = pieceTypeOf(piece)

		board[square] = piece
		pieces[color][pieceType] = addSquare(square, pieces[color][pieceType])
	}

	fun remove(square: Square): Piece {
		val piece = board[square]
		val color = pieceColorOf(piece)
		val pieceType = pieceTypeOf(piece)

		board[square] = NO_PIECE
		pieces[color][pieceType] = removeSquare(square, pieces[color][pieceType])

		return piece
	}

	fun setCastlingRight(castling: Castling) {
		castlingRights = castlingRights or castling
	}

	private fun clearCastling(square: Square) {
		when (square) {
			A1 -> castlingRights = castlingRights and WHITE_QUEENSIDE.inv()
			A8 -> castlingRights = castlingRights and BLACK_QUEENSIDE.inv()
			H1 -> castlingRights = castlingRights and WHITE_KINGSIDE.inv()
			H8 -> castlingRights = castlingRights and BLACK_KINGSIDE.inv()
			E1 -> castlingRights = castlingRights and (WHITE_KINGSIDE or WHITE_QUEENSIDE).inv()
			E8 -> castlingRights = castlingRights and (BLACK_KINGSIDE or BLACK_QUEENSIDE).inv()
		}
	}

	override fun equals(other: Any?): Boolean {
		if (this === other) return true
		if (other !is Position) return false

		return this.activeColor == other.activeColor
			&& this.castlingRights == other.castlingRights
			&& this.enPassantSquare == other.enPassantSquare
			&& this.halfmoveClock == other.halfmoveClock
			&& this.halfmoveNumber == other.halfmoveNumber
			&& this.board contentDeepEquals other.board
			&& this.pieces contentDeepEquals other.pieces
	}
}

private class StateList {
	var size: Int = 0
	val entries: Array<StateEntry> = Array(MAX_PLY) { StateEntry() }
}

private class StateEntry {
	var castlingRights: Castling = NO_CASTLING
	var enPassantSquare: Square = NO_SQUARE
	var halfmoveClock: Int = 0
}

File: pulse-go/internal/pulse/engine/position.go
/*
 * Copyright 2013-2024 Phokham Nonava
 *
 * Use of this source code is governed by the MIT license that can be
 * found in the LICENSE file.
 */

package engine

import (
	"fmt"
)

func NewPosition() *Position {
	position := &Position{
		ActiveColor:     NoColor,
		CastlingRights:  NoCastling,
		EnPassantSquare: NoSquare,
		HalfmoveClock:   0,
		HalfmoveNumber:  0,
		board:           [squaresMaxValue]Piece{},
		pieces:          [len(Colors)][len(PieceTypes)]bitboard{},
		stateList: stateList{
			size:    0,
			entries: [MaxPly]stateEntry{},
		},
	}
	for sq := range position.board {
		position.board[sq] = NoPiece
	}
	for col := range position.pieces {
		for pt := range position.pieces[col] {
			position.pieces[col][pt] = bitboard(0)
		}
	}
	for i := range position.stateList.entries {
		position.stateList.entries[i] = stateEntry{
			castlingRights:  NoCastling,
			enPassantSquare: NoSquare,
			halfmoveClock:   0,
		}
	}
	return position
}

type Position struct {
	ActiveColor     Color
	CastlingRights  Castling
	EnPassantSquare Square
	HalfmoveClock   int
	HalfmoveNumber  int

	board  [squaresMaxValue]Piece
	pieces [len(Colors)][len(PieceTypes)]bitboard

	stateList stateList
}

func (p *Position) MakeMove(m Move) {
	p.saveState()

	mt := MoveTypeOf(m)
	originSq := OriginSquareOf(m)
	targetSq := TargetSquareOf(m)
	originPc := OriginPieceOf(m)
	originCol := PieceColorOf(originPc)
	targetPc := TargetPieceOf(m)

	if targetPc != NoPiece {
		captureSq := targetSq
		if mt == EnPassantMove {
			oppositeDir := pawnMoveDirections[OppositeOf(originCol)]
			captureSq += oppositeDir
		}
		p.Remove(captureSq)
		p.clearCastling(captureSq)
	}

	p.Remove(originSq)
	if mt == PawnPromotionMove {
		p.Put(PieceOf(originCol, PromotionOf(m)), targetSq)
	} else {
		p.Put(originPc, targetSq)
	}

	if mt == CastlingMove {
		switch targetSq {
		case G1:
			p.Put(p.Remove(H1), F1)
		case C1:
			p.Put(p.Remove(A1), D1)
		case G8:
			p.Put(p.Remove(H8), F8)
		case C8:
			p.Put(p.Remove(A8), D8)
		default:
			panic(fmt.Sprintf("Invalid target square: %v", targetSq))
		}
	}

	p.clearCastling(originSq)

	if mt == PawnDoubleMove {
		oppositeDir := pawnMoveDirections[OppositeOf(originCol)]
		p.EnPassantSquare = targetSq + oppositeDir
	} else {
		p.EnPassantSquare = NoSquare
	}

	p.ActiveColor = OppositeOf(p.ActiveColor)

	if PieceTypeOf(originPc) == Pawn || targetPc != NoPiece {
		p.HalfmoveClock = 0
	} else {
		p.HalfmoveClock++
	}

	p.HalfmoveNumber++
}

func (p *Position) UndoMove(m Move) {
	mt := MoveTypeOf(m)
	originSq := OriginSquareOf(m)
	targetSq := TargetSquareOf(m)
	originPc := OriginPieceOf(m)
	originCol := PieceColorOf(originPc)
	targetPc := TargetPieceOf(m)

	p.HalfmoveNumber--

	p.ActiveColor = OppositeOf(p.ActiveColor)

	if mt == CastlingMove {
		switch targetSq {
		case G1:
			p.Put(p.Remove(F1), H1)
		case C1:
			p.Put(p.Remove(D1), A1)
		case G8:
			p.Put(p.Remove(F8), H8)
		case C8:
			p.Put(p.Remove(D8), A8)
		default:
			panic(fmt.Sprintf("Invalid target square: %v", targetSq))
		}
	}

	p.Remove(targetSq)
	p.Put(originPc, originSq)

	if targetPc != NoPiece {
		captureSq := targetSq
		if mt == EnPassantMove {
			oppositeDir := pawnMoveDirections[OppositeOf(originCol)]
			captureSq += oppositeDir
		}
		p.Put(targetPc, captureSq)
	}

	p.restoreState()
}

func (p *Position) saveState() {
	entry := &p.stateList.entries[p.stateList.size]
	entry.castlingRights = p.CastlingRights
	entry.enPassantSquare = p.EnPassantSquare
	entry.halfmoveClock = p.HalfmoveClock
	p.stateList.size++
}

func (p *Position) restoreState() {
	p.stateList.size--
	entry := &p.stateList.entries[p.stateList.size]
	p.CastlingRights = entry.castlingRights
	p.EnPassantSquare = entry.enPassantSquare
	p.HalfmoveClock = entry.halfmoveClock
}

func (p *Position) Get(sq Square) Piece {
	return p.board[sq]
}

func (p *Position) Put(pc Piece, sq Square) {
	col := PieceColorOf(pc)
	pt := PieceTypeOf(pc)

	p.board[sq] = pc
	p.pieces[col][pt] = addSquare(sq, p.pieces[col][pt])
}

func (p *Position) Remove(sq Square) Piece {
	pc := p.board[sq]
	col := PieceColorOf(pc)
	pt := PieceTypeOf(pc)

	p.board[sq] = NoPiece
	p.pieces[col][pt] = removeSquare(sq, p.pieces[col][pt])

	return pc
}

func (p *Position) SetCastlingRight(cast Castling) {
	p.CastlingRights |= cast
}

func (p *Position) clearCastling(sq Square) {
	switch sq {
	case A1:
		p.CastlingRights = p.CastlingRights & ^WhiteQueenside
	case A8:
		p.CastlingRights = p.CastlingRights & ^BlackQueenside
	case H1:
		p.CastlingRights = p.CastlingRights & ^WhiteKingside
	case H8:
		p.CastlingRights = p.CastlingRights & ^BlackKingside
	case E1:
		p.CastlingRights = p.CastlingRights & ^(WhiteKingside | WhiteQueenside)
	case E8:
		p.CastlingRights = p.CastlingRights & ^(BlackKingside | BlackQueenside)
	}
}

type stateList struct {
	size    int
	entries [MaxPly]stateEntry
}

type stateEntry struct {
	castlingRights  Castling
	enPassantSquare Square
	halfmoveClock   int
}
